Ceph FS fscrypt support

irq0 opened the pull request:
Add Ceph FS fscrypt support. Supports volumes, snapshots and clones.

Follow up to #3310 (adds fscrypt integration) and second in series towards #1563

Fixes: #1563 

# Usage Requirements

- Kernel with Ceph FS fscrypt support
   -   git://git.kernel.org/pub/scm/linux/kernel/git/jlayton/linux.git branch ceph-fscrypt
   -   or <https://github.com/ceph/ceph-client.git> branch wip-fscrypt
-   <https://github.com/google/fscrypt#runtime-dependencies>
-   Ceph MDS support
    -   v17: volumes supported
    -   main branch: +snapshots
    -   main branch+<https://github.com/ceph/ceph/pull/48410>: +clones


# End-to-end Testing

Testing on minikube requires a custom ISO with the latest Ceph FS patches. To test snapshots and clones a main branch build of Ceph is needed as well.

## Custom Minikube ISO

Use patches from <https://github.com/irq0/minikube/tree/custom-ceph-fscrypt-kernel> on top of minikube to build an ISO with the current Ceph FS development kernel.

Build summary:

```
make
make buildroot-image
make out/minikube-x86_64.iso
```

Then use minikube.sh with MINIKUBE_ISO_URL="file://path of minikube.iso"

## Use Ceph Main Branch Images
Either use `quay.io/ceph/daemon:latest-main-devel` or build using ceph-containers project.

## Build

In <https://github.com/ceph/ceph-container.git> clone:

```
eval $(minikube docker-env)  # will build inside minikube VM
make BASEOS_REGISTRY=quay.io/centos BASEOS_TAG=stream8 CEPH_DEVEL=true FLAVORS="main,centos,8" build
```

Optional: patch image with <https://github.com/ceph/ceph/pull/48410>

Commit:
cephfs: fscrypt encryption support

Add Ceph FS fscrypt support, similar to the RBD/ext4 fscrypt
integration. Supports encrypted PVCs, snapshots and clones.

Requires kernel and Ceph MDS support that is currently not in any
stable release.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
deploy: Add KMS configuration to Ceph FS

Adds necessary KMS configuration based on the RBD configuration to use
Ceph FS with fscrypt

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
examples: Ceph FS fscrypt / KMS additions

Add encryption configuration to Ceph FS examples

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Add Ceph FS fscrypt validation helper

Add e2e helper to verify encrypted Ceph FS. Verify file's
ceph.fscrypt.auth attribute and KMS password creation / removal.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Deploy vault as part of the Ceph FS suite

Always deploy Vault as part of the the Ceph FS test suite.

Required by:
 - fscrypt tests using any vault KMS type.
 - Configuration in deploy/cephfs/*.yaml via the
   ceph-csi-encryption-kms-config config map created during deployVault()

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: add basic PVC Ceph FS fscrypt tests

Test storage class, pvc and app bind of an fscrypt encrypted Ceph FS
with secrets metadata, vault, vault tokens and vault tenant KMS.

Tests are based on the RBD block/file encryption tests.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: add PVC-PVC clone Ceph FS fscrypt tests

Note: Feature fixed https://tracker.ceph.com/issues/57641

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: add cephfs fscrypt snapshot volume test

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Add test-cephfs-fscrypt flag

Add flag to default disable Ceph FS fscrypt tests, as they require a
custom minikube ISO

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
Test last push addresses the linter errors.

The failure in ci/centos/upgrade-tests-cephfs is interesting:

The updated deploy/cephfs/* files now have the ceph-csi-encryption-kms-config configmaps attached as volumes (like their RBD counterparts). My read is that the update tests fail, because they are not present. What is the best course of action here? Add the encryption related configmaps to the update path? Fork the file into regular and encryption enabled?

Madhu-1 commented:
> ceph-csi-encryption-kms-config configmaps attached as volumes (like their RBD counterparts). My read is that the update tests fail, because they are not present.

You can set https://github.com/ceph/ceph-csi/blob/devel/e2e/cephfs.go#L73 to true so that it will continue if the file is not present

Commit:
cephfs: nolint:gocyclo NewVolumeOptions, NewVolumeOptionsFromVolID

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
e2e: Deploy vault as part of the Ceph FS upgrade suite

Deploy vault, analogue to the RBD upgrade suite to have the
ceph-csi-encryption-kms-config map ready when dependent
deployments/daemonsets are created.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
> > ceph-csi-encryption-kms-config configmaps attached as volumes (like their RBD counterparts). My read is that the update tests fail, because they are not present.
> 
> You can set https://github.com/ceph/ceph-csi/blob/devel/e2e/cephfs.go#L73 to true so that it will continue if the file is not present

Adding deployVault (just like the RBD suites have) to the tests did the trick. It configures the ceph-csi-encryption-kms-config required by the cephfs provisioner / nodeserver

irq0 commented:
/retest ci/centos/upgrade-tests-cephfs

irq0 commented:
/retest ci/centos/mini-e2e/k8s-1.22
/retest ci/centos/mini-e2e-helm/k8s-1.24
/retest ci/centos/mini-e2e-helm/k8s-1.23
/retest ci/centos/k8s-e2e-external-storage/1.22

irq0 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.24

irq0 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

irq0 commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

irq0 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

irq0 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

Madhu-1 commented on deploy/cephfs/kubernetes/csi-nodeplugin-rbac.yaml:
```diff
@@ -3,3 +3,46 @@ apiVersion: v1
 kind: ServiceAccount
 metadata:
   name: cephfs-csi-nodeplugin
+  namespace: default
```
use `namespace: {{ .Release.Namespace }}` here

irq0 replied:
The RBD analogue doesn't have that construct. It is for use in the helm templates, not?

Madhu-1 commented on deploy/cephfs/kubernetes/csi-nodeplugin-rbac.yaml:
```diff
@@ -3,3 +3,46 @@ apiVersion: v1
+metadata:
+  name: cephfs-csi-nodeplugin
+rules:
+  - apiGroups: [""]
+    resources: ["nodes"]
+    verbs: ["get"]
```
This is not required for cephfs encryption right?

irq0 replied:
Removed

Madhu-1 commented on deploy/cephfs/kubernetes/csi-nodeplugin-rbac.yaml:
```diff
@@ -3,3 +3,46 @@ apiVersion: v1
+    resources: ["serviceaccounts"]
+    verbs: ["get"]
+  - apiGroups: [""]
+    resources: ["persistentvolumes"]
+    verbs: ["get"]
+  - apiGroups: ["storage.k8s.io"]
+    resources: ["volumeattachments"]
+    verbs: ["list", "get"]
```
This is also not required for cephfs encryption right

irq0 replied:
Removed

Madhu-1 commented on deploy/cephfs/kubernetes/csi-provisioner-rbac.yaml:
```diff
@@ -68,6 +81,9 @@ metadata:
   namespace: default
   name: cephfs-external-provisioner-cfg
 rules:
+  - apiGroups: [""]
+    resources: ["configmaps"]
+    verbs: ["get", "list", "watch", "create", "update", "delete"]
```
Why do we need to have create,update, and delete access?

irq0 replied:
Removed

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -541,7 +556,19 @@ func (cs *ControllerServer) cleanUpBackingVolume(
+		// fscrypt on the volume that is going to be deleted anyway.
+		log.DebugLog(ctx, "cephfs: going to remove DEK for integrated store %q (fscrypt)", volOptions.Encryption.GetID())
+		if err := volOptions.Encryption.RemoveDEK(volID.VolumeID); err != nil {
+			log.WarningLog(ctx, "failed to clean the passphrase for volume %+v (file encryption): %s",
```
`%q` instead of `%+v`

irq0 replied:
:heavy_check_mark:

Madhu-1 commented on internal/cephfs/controllerserver.go:
```diff
@@ -541,7 +556,19 @@ func (cs *ControllerServer) cleanUpBackingVolume(
+		// Only remove DEK when the KMS stores it itself. On
+		// GetSecret enabled KMS the DEKs are stored by
+		// fscrypt on the volume that is going to be deleted anyway.
+		log.DebugLog(ctx, "cephfs: going to remove DEK for integrated store %q (fscrypt)", volOptions.Encryption.GetID())
```
no need to use `cephfs` in debug log

irq0 replied:
:+1:

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -118,6 +119,40 @@ func validateSnapshotBackedVolCapability(volCap *csi.VolumeCapability) error {
 	return nil
 }
 
+// maybeUnlockFileEncryption unlocks fscrypt on stagingTargetPath, iff volOptions enable encryption.
```
`iff` to `if`

irq0 replied:
:heavy_check_mark:

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -118,6 +119,40 @@ func validateSnapshotBackedVolCapability(volCap *csi.VolumeCapability) error {
+	return nil
+}
+
+// maybeInitializeFileEncryption initializes KMS and node specifics, iff volContext enables encryption.
```
`iff` to `if`

irq0 replied:
:heavy_check_mark:

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -118,6 +119,40 @@ func validateSnapshotBackedVolCapability(volCap *csi.VolumeCapability) error {
+) error {
+	if volOptions.IsEncrypted() {
+		if _, isFuse := mnt.(*mounter.FuseMounter); isFuse {
+			return fmt.Errorf("FUSE mounter does not support encryption")
```
use errors.New instead of `fmt.Errof` as formating is not required.

irq0 replied:
:+1:

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -118,6 +119,40 @@ func validateSnapshotBackedVolCapability(volCap *csi.VolumeCapability) error {
+		if _, isFuse := mnt.(*mounter.FuseMounter); isFuse {
+			return fmt.Errorf("FUSE mounter does not support encryption")
+		}
+		if err := fscrypt.InitializeNode(ctx); err != nil {
+			return err
+		}
```
```suggestion
return fscrypt.InitializeNode(ctx)
```

irq0 replied:
:+1:

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -118,6 +119,40 @@ func validateSnapshotBackedVolCapability(volCap *csi.VolumeCapability) error {
 }
 
+// maybeUnlockFileEncryption unlocks fscrypt on stagingTargetPath, iff volOptions enable encryption.
+func maybeUnlockFileEncryption(
```
Any specific reason to start with `maybe` for both functions?

irq0 replied:
Both functions only do things if encryption is enabled. The 'maybe' tried to convey that the function may not do anything at all.

Madhu-1 reviewed (commented):
Changes look good, with some minor nits. Can you point to some document that can help us to test this feature on local systems (ceph image, kernel version to use), etc?

irq0 commented:
> Can you point to some document that can help us to test this feature on local systems (ceph image, kernel version to use), etc?

I uploaded a minikube ISO with the current development Ceph FS fscrypt kernel that might be helpful: https://github.com/irq0/minikube/releases



You can test most changes by setting the following when calling scripts/minikube.sh

```
ROOK_CEPH_CLUSTER_IMAGE=quay.io/ceph/daemon:latest-main-devel
MINIKUBE_ISO_URL="https://github.com/irq0/minikube/releases/download/2022-11-08-master%2Bfscrypt-kernel/minikube-amd64.iso"
```

My helper scripts are in https://github.com/irq0/dev-ceph-csi-fscrypt-config
Calling kube-deploy-cephcsi from there sets up ceph csi and a deployment using an already set up rook.

Commit:
deploy: Remove unnecessary RBAC permissions

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

irq0 commented:
Push: Rebase to devel and changes for the comments above

nixpanic commented on e2e/cephfs.go:
```diff
@@ -400,6 +400,67 @@ var _ = Describe(cephfsType, func() {
+				for kmsID, kmsConf := range kmsToTest {
+					kmsID := kmsID
+					kmsConf := kmsConf
+					By("create a storageclass with pool and an encrypted PVC then bind it to an app with "+kmsID, func() {
```
I'm not happy with the extreme indention here... But I also don't immediately see how this can be cleanly simplified.

nixpanic commented on internal/cephfs/store/volumeoptions.go:
```diff
@@ -210,6 +210,7 @@ func fmtBackingSnapshotOptionMismatch(optName, expected, actual string) error {
 
 // NewVolumeOptions generates a new instance of volumeOptions from the provided
 // CSI request parameters.
+// nolint:gocyclo,cyclop // TODO: reduce complexity
```
this probably should be squashed in the commit that added the `InitKMS()` and/or other calls?

nixpanic commented on deploy/cephfs/kubernetes/csi-nodeplugin-rbac.yaml:
```diff
@@ -10,10 +10,6 @@ apiVersion: rbac.authorization.k8s.io/v1
 metadata:
   name: cephfs-csi-nodeplugin
 rules:
```
This is a nice cleanup, and could have been a PR by itself.

nixpanic reviewed (approved):
Just a few nits, nothing that requires correcting in my opinion. I'm a little sad I can't easily test this out. As long as it does not break anything, we should be good for now.

Ideally we start running weekly tests with a non-standard kernel config (minkube VM?) and Ceph release that contains support for fscrypt.

irq0 commented:
> Just a few nits, nothing that requires correcting in my opinion. I'm a little sad I can't easily test this out. As long as it does not break anything, we should be good for now.

Thanks! Me too. It would be awesome to get at least some CI coverage of the Ceph FS fscrypt path.

> 
> Ideally we start running weekly tests with a non-standard kernel config (minkube VM?) and Ceph release that contains support for fscrypt.

A minikube VM is probably the best option. Comment https://github.com/ceph/ceph-csi/pull/3460#issuecomment-1307347841 has some details on how.

I think a custom ISO is the only way to go for the foreseeable future, as minikube kernels tend to be behind and fscrypt kernel support isn't (fully) merged yet.

Madhu-1 commented:
/test ci/centos/upgrade-tests-cephfs

humblec commented:
@irq0 thanks for pulling it  ! lgtm..  its also good to add a user facing documentation for CephFS Fscrypt. It could be a kind of placeholder doc with disclaimers ( required kernel support..etc) too .

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.24

nixpanic commented:
@mergifyio requeue
